deploy/containers/credhub_setup/auth.go

```diff
@@ -0,0 +1,120 @@
+package main
+
+// This file contains code to deal with UAA / authentication.
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"time"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+func makeHTTPClientWithCA(serverName string, caCert []byte) (*http.Client, error) {
+	certPool, err := x509.SystemCertPool()
+	if err != nil {
+		return nil, fmt.Errorf("could not get system cert pool: %w", err)
+	}
+	ok := certPool.AppendCertsFromPEM(caCert)
+	if !ok {
+		fmt.Printf("Warning: failed to add CA certs for %s\n", serverName)
+	}
+	return &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs:    certPool,
+				ServerName: serverName,
+			},
+		},
+		Timeout: 60 * time.Second,
+	}, nil
+}
+
+// authenticate with UAA, returning the access token and refresh token.
+func authenticate(ctx context.Context, clientID, clientSecret string) (*http.Client, error) {
+	link, err := getCCLinkData(ctx)
+	if err != nil {
+		return nil, err
+	}
+	ccClient, err := makeHTTPClientWithCA(link.CC.InternalServiceHostname, []byte(link.CC.PublicTLS.CACert))
```

Comment:
`makeHTTPClientWithCA` is a hidden dependency. An instance of `HttpClient` should be passed as a dependency to the constructor of the struct. What we accept should be an interface that can be mocked in order to unit test this behaviour.